File: worker/lib/providers.js
/**
 * Multi-provider adapter for LLM APIs.
 *
 * Soporta:
 *   - anthropic  (Claude Haiku/Sonnet/Opus)   — api.anthropic.com
 *   - openai     (GPT-5 / GPT-5 mini / nano)  — api.openai.com
 *   - google     (Gemini 2.5 Flash/Pro)       — generativelanguage.googleapis.com
 *
 * Uso:
 *   import { chat, PROVIDERS, PRICING } from './lib/providers.js';
 *
 *   const result = await chat(env, {
 *     provider: 'google',
 *     model:    'gemini-2.5-flash',
 *     system:   'Eres un asistente...',
 *     messages: [{role: 'user', content: '...'}],
 *     maxTokens: 1024,
 *     jsonMode: true,         // opcional, fuerza JSON
 *     temperature: 0.0
 *   });
 *   // → { text, usage: {inputTokens, outputTokens}, model, costUsd, provider }
 *
 * Las claves se leen del env: ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY
 */

// ============================================================
// PRICING (USD per million tokens, mayo 2026)
// ============================================================
export const PRICING = {
  // Anthropic
  'claude-haiku-4-5':       { in: 1.00,  out: 5.00  },
  'claude-haiku-4-5-20251001': { in: 1.00, out: 5.00 },
  'claude-sonnet-4-6':      { in: 3.00,  out: 15.00 },
  'claude-opus-4-7':        { in: 5.00,  out: 25.00 },
  'claude-opus-4-6':        { in: 5.00,  out: 25.00 },

  // OpenAI (estimados — verifica platform.openai.com/pricing)
  'gpt-5':                  { in: 3.00,  out: 12.00 },
  'gpt-5-mini':             { in: 0.25, out: 2.00 },
  'gpt-5-nano':             { in: 0.10, out: 0.40 },
  'gpt-4o':                 { in: 2.50, out: 10.00 },
  'gpt-4o-mini':            { in: 0.15, out: 0.60 },

  // Google (estimados — verifica ai.google.dev/pricing)
  'gemini-2.5-flash':       { in: 0.10, out: 0.40 },
  'gemini-2.5-flash-lite':  { in: 0.075, out: 0.30 },
  'gemini-2.5-pro':         { in: 1.25, out: 5.00 },
  'gemini-2.0-flash':       { in: 0.10, out: 0.40 },
};

// ============================================================
// PROVIDERS catalog
// ============================================================
export const PROVIDERS = {
  anthropic: {
    label: 'Anthropic Claude',
    defaultModel: 'claude-sonnet-4-6',
    cheapModel:   'claude-haiku-4-5',
    models: ['claude-haiku-4-5', 'claude-sonnet-4-6', 'claude-opus-4-7'],
    envKey: 'ANTHROPIC_API_KEY',
  },
  openai: {
    label: 'OpenAI GPT',
    defaultModel: 'gpt-5',
    cheapModel:   'gpt-5-mini',
    models: ['gpt-5-nano', 'gpt-5-mini', 'gpt-5', 'gpt-4o', 'gpt-4o-mini'],
    envKey: 'OPENAI_API_KEY',
  },
  google: {
    label: 'Google Gemini',
    defaultModel: 'gemini-2.5-pro',
    cheapModel:   'gemini-2.5-flash',
    models: ['gemini-2.5-flash-lite', 'gemini-2.5-flash', 'gemini-2.5-pro'],
    envKey: 'GOOGLE_API_KEY',
  },
};

export function costUsd(model, inputTokens, outputTokens) {
  const p = PRICING[model];
  if (!p) return null;
  return (inputTokens * p.in + outputTokens * p.out) / 1_000_000;
}

// ============================================================
// MAIN: chat()
// ============================================================
export async function chat(env, opts) {
  const provider = opts.provider || 'anthropic';
  const model = opts.model || PROVIDERS[provider]?.defaultModel;
  if (!model) throw new Error(`Unknown provider ${provider}`);

  const apiKey = env[PROVIDERS[provider].envKey];
  if (!apiKey) {
    throw new Error(`Missing ${PROVIDERS[provider].envKey} in env`);
  }

  let result;
  switch (provider) {
    case 'anthropic': result = await anthropicChat(apiKey, model, opts); break;
    case 'openai':    result = await openaiChat(apiKey, model, opts); break;
    case 'google':    result = await googleChat(apiKey, model, opts); break;
    default: throw new Error(`Unknown provider ${provider}`);
  }
  result.provider = provider;
  result.model = model;
  result.costUsd = costUsd(model, result.usage.inputTokens, result.usage.outputTokens);
  return result;
}

// ============================================================
// ANTHROPIC adapter
// ============================================================
async function anthropicChat(apiKey, model, opts) {
  const body = {
    model,
    max_tokens: opts.maxTokens || 1024,
    messages: opts.messages,
  };
  if (opts.system) body.system = opts.system;
  if (opts.temperature !== undefined) body.temperature = opts.temperature;

  const res = await fetch('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
      'content-type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const t = await res.text();
    throw new Error(`Anthropic ${res.status}: ${t.slice(0, 300)}`);
  }
  const data = await res.json();
  return {
    text: data.content?.[0]?.text || '',
    usage: {
      inputTokens: data.usage?.input_tokens || 0,
      outputTokens: data.usage?.output_tokens || 0,
    },
    finishReason: data.stop_reason || 'unknown',
    raw: data,
  };
}

// ============================================================
// OPENAI adapter (Chat Completions format)
// ============================================================
async function openaiChat(apiKey, model, opts) {
  const messages = [];
  if (opts.system) messages.push({ role: 'system', content: opts.system });
  for (const m of opts.messages) messages.push({ role: m.role, content: m.content });

  const body = {
    model,
    messages,
    max_completion_tokens: opts.maxTokens || 1024,
  };
  if (opts.temperature !== undefined) body.temperature = opts.temperature;
  if (opts.jsonMode) body.response_format = { type: 'json_object' };

  const res = await fetch('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const t = await res.text();
    throw new Error(`OpenAI ${res.status}: ${t.slice(0, 300)}`);
  }
  const data = await res.json();
  return {
    text: data.choices?.[0]?.message?.content || '',
    usage: {
      inputTokens: data.usage?.prompt_tokens || 0,
      outputTokens: data.usage?.completion_tokens || 0,
    },
    finishReason: data.choices?.[0]?.finish_reason || 'unknown',
    raw: data,
  };
}

// ============================================================
// GOOGLE GEMINI adapter
// ============================================================
async function googleChat(apiKey, model, opts) {
  // Gemini usa "contents" array con roles user/model (no assistant)
  const contents = opts.messages.map(m => ({
    role: m.role === 'assistant' ? 'model' : m.role,
    parts: [{ text: m.content }],
  }));

  const body = {
    contents,
    generationConfig: {
      maxOutputTokens: opts.maxTokens || 1024,
    },
  };
  if (opts.system) {
    body.systemInstruction = { parts: [{ text: opts.system }] };
  }
  if (opts.temperature !== undefined) {
    body.generationConfig.temperature = opts.temperature;
  }
  if (opts.jsonMode) {
    body.generationConfig.responseMimeType = 'application/json';
  }

  const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const t = await res.text();
    throw new Error(`Google ${res.status}: ${t.slice(0, 300)}`);
  }
  const data = await res.json();
  const candidate = data.candidates?.[0];
  const text = candidate?.content?.parts?.map(p => p.text).join('') || '';
  return {
    text,
    usage: {
      inputTokens: data.usageMetadata?.promptTokenCount || 0,
      outputTokens: data.usageMetadata?.candidatesTokenCount || 0,
    },
    finishReason: candidate?.finishReason || 'unknown',
    raw: data,
  };
}

// ============================================================
// Helper: resolve provider+model from optional request body + env defaults
// ============================================================
export function resolveProvider(env, body, fallback) {
  const provider = body?.provider || env.DEFAULT_PROVIDER || fallback.provider;
  const model = body?.model || env.DEFAULT_MODEL || fallback.model;
  return { provider, model };
}

File: scripts/providers.py
#!/usr/bin/env python3
"""
Multi-provider adapter para Python.

Mismo concepto que worker/lib/providers.js, pero para usar desde scripts/GitHub Actions.

Uso:
    from providers import chat, PROVIDERS

    result = chat(
        provider='google',
        model='gemini-2.5-flash',
        system='Eres un asistente...',
        messages=[{'role':'user','content':'...'}],
        max_tokens=600,
        json_mode=True
    )
    # → {'text': str, 'usage': {'input_tokens', 'output_tokens'}, 'model', 'cost_usd', 'provider'}

Env vars: ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY
"""

import os
import json
from typing import Optional
import requests

PRICING = {
    # Anthropic
    'claude-haiku-4-5':       {'in': 1.00, 'out': 5.00},
    'claude-haiku-4-5-20251001': {'in': 1.00, 'out': 5.00},
    'claude-sonnet-4-6':      {'in': 3.00, 'out': 15.00},
    'claude-opus-4-7':        {'in': 5.00, 'out': 25.00},
    # OpenAI
    'gpt-5':                  {'in': 3.00, 'out': 12.00},
    'gpt-5-mini':             {'in': 0.25, 'out': 2.00},
    'gpt-5-nano':             {'in': 0.10, 'out': 0.40},
    'gpt-4o':                 {'in': 2.50, 'out': 10.00},
    'gpt-4o-mini':            {'in': 0.15, 'out': 0.60},
    # Google
    'gemini-2.5-flash':       {'in': 0.10, 'out': 0.40},
    'gemini-2.5-flash-lite':  {'in': 0.075, 'out': 0.30},
    'gemini-2.5-pro':         {'in': 1.25, 'out': 5.00},
}

PROVIDERS = {
    'anthropic': {
        'label': 'Anthropic Claude',
        'default_model': 'claude-sonnet-4-6',
        'cheap_model':   'claude-haiku-4-5',
        'env_key': 'ANTHROPIC_API_KEY',
    },
    'openai': {
        'label': 'OpenAI GPT',
        'default_model': 'gpt-5',
        'cheap_model':   'gpt-5-mini',
        'env_key': 'OPENAI_API_KEY',
    },
    'google': {
        'label': 'Google Gemini',
        'default_model': 'gemini-2.5-pro',
        'cheap_model':   'gemini-2.5-flash',
        'env_key': 'GOOGLE_API_KEY',
    },
}


def cost_usd(model: str, input_tokens: int, output_tokens: int) -> Optional[float]:
    p = PRICING.get(model)
    if not p:
        return None
    return (input_tokens * p['in'] + output_tokens * p['out']) / 1_000_000


def chat(provider='anthropic', model=None, system=None, messages=None,
         max_tokens=1024, temperature=None, json_mode=False, timeout=60):
    """Llamada unificada a cualquiera de los 3 proveedores."""
    if provider not in PROVIDERS:
        raise ValueError(f"Unknown provider {provider}")
    if model is None:
        model = PROVIDERS[provider]['default_model']
    env_key = PROVIDERS[provider]['env_key']
    api_key = os.environ.get(env_key)
    if not api_key:
        raise RuntimeError(f"Missing env var {env_key}")

    if provider == 'anthropic':
        result = _anthropic_chat(api_key, model, system, messages, max_tokens, temperature, timeout)
    elif provider == 'openai':
        result = _openai_chat(api_key, model, system, messages, max_tokens, temperature, json_mode, timeout)
    elif provider == 'google':
        result = _google_chat(api_key, model, system, messages, max_tokens, temperature, json_mode, timeout)

    result['provider'] = provider
    result['model'] = model
    result['cost_usd'] = cost_usd(model, result['usage']['input_tokens'], result['usage']['output_tokens'])
    return result


def _anthropic_chat(api_key, model, system, messages, max_tokens, temperature, timeout):
    body = {'model': model, 'max_tokens': max_tokens, 'messages': messages}
    if system: body['system'] = system
    if temperature is not None: body['temperature'] = temperature
    r = requests.post(
        'https://api.anthropic.com/v1/messages',
        headers={'x-api-key': api_key, 'anthropic-version': '2023-06-01', 'content-type': 'application/json'},
        json=body, timeout=timeout
    )
    r.raise_for_status()
    data = r.json()
    return {
        'text': (data.get('content') or [{}])[0].get('text', ''),
        'usage': {
            'input_tokens': data.get('usage', {}).get('input_tokens', 0),
            'output_tokens': data.get('usage', {}).get('output_tokens', 0),
        },
        'finish_reason': data.get('stop_reason', 'unknown'),
    }


def _openai_chat(api_key, model, system, messages, max_tokens, temperature, json_mode, timeout):
    msgs = []
    if system: msgs.append({'role': 'system', 'content': system})
    msgs.extend(messages)
    body = {'model': model, 'messages': msgs, 'max_completion_tokens': max_tokens}
    if temperature is not None: body['temperature'] = temperature
    if json_mode: body['response_format'] = {'type': 'json_object'}
    r = requests.post(
        'https://api.openai.com/v1/chat/completions',
        headers={'Authorization': f'Bearer {api_key}', 'Content-Type': 'application/json'},
        json=body, timeout=timeout
    )
    r.raise_for_status()
    data = r.json()
    choice = (data.get('choices') or [{}])[0]
    return {
        'text': choice.get('message', {}).get('content', '') or '',
        'usage': {
            'input_tokens': data.get('usage', {}).get('prompt_tokens', 0),
            'output_tokens': data.get('usage', {}).get('completion_tokens', 0),
        },
        'finish_reason': choice.get('finish_reason', 'unknown'),
    }


def _google_chat(api_key, model, system, messages, max_tokens, temperature, json_mode, timeout):
    contents = [{
        'role': 'model' if m['role'] == 'assistant' else m['role'],
        'parts': [{'text': m['content']}]
    } for m in messages]
    body = {
        'contents': contents,
        'generationConfig': {'maxOutputTokens': max_tokens}
    }
    if system:
        body['systemInstruction'] = {'parts': [{'text': system}]}
    if temperature is not None:
        body['generationConfig']['temperature'] = temperature
    if json_mode:
        body['generationConfig']['responseMimeType'] = 'application/json'
    url = f'https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={api_key}'
    r = requests.post(url, json=body, timeout=timeout)
    r.raise_for_status()
    data = r.json()
    cand = (data.get('candidates') or [{}])[0]
    text = ''.join(p.get('text', '') for p in (cand.get('content') or {}).get('parts', []))
    meta = data.get('usageMetadata', {})
    return {
        'text': text,
        'usage': {
            'input_tokens': meta.get('promptTokenCount', 0),
            'output_tokens': meta.get('candidatesTokenCount', 0),
        },
        'finish_reason': cand.get('finishReason', 'unknown'),
    }
